### PR TITLE
Handle server errors

### PR DESCRIPTION
**Pull request by salvafern**
Hello there!

Here we added tryCatch in `emodnet_init_wfs_client()` to handle errors related to the server / user internet connection. This way we can separate package errors from server errors. 

When a client fails to be created, the following output is shown in the console:

```
#> x WFS client creation failed.
#> i Service: 'https://demo.geo-solutions.it/geoserver/ows'
#> i HTTP Status: Server error: (500) Internal Server Error
#>
#> Browse the EMODnet OGC monitor?
#>
#> 1: Definitely
#> 2: No
#> 3: Absolutely not
```

### New functions

There are two new functions. I think there's no need to document them with rOxygen:
* `perform_http_request()`: Checks if there is internet connect and performs a GetCapabilities request as HTTP
* `check_service()`: reads the request and provides feedback: notices if it was a bad request (status != 200 or 2**) and asks the user if wants to browse the emodnet monitor tool  (monitor.emodnet.eu), then returns NULL. 

However, If the request was correct (status = 200 or 2**), then raises an error. As this is going to be used in tryCatch, it will only be triggered if an error arises when creating the WFS client. if the status of the server is correct then we are in an unexpected situation and it might be a problem of the package
   
### Testing

Some new tests were added here: https://github.com/EMODnet/EMODnetWFS/blob/95f20ae75a92a520fc4cb7f7e5bd2c96276431d0/tests/testthat/test-client.R#L17

To make the tests independent from internet connection and the actual status of an external WFS service, I added [webmockr](https://docs.ropensci.org/webmockr/) to mock an HTTP response that works and one that doesn't. The tests are performed on these two mocked requests.

@maelle Could you have a look at this tests? I also tried using vcr but webmockr seemed more relevant for this case. I am very new to these two packages so any feedback is very much welcome :)

### CODE REVIEW

**Comment by salvafern** [comment]
Notice I decided to not check the status of the services by requesting info from the monitor.emodnet.eu tool. Instead, I check directly the http status of the GetCapabilities WFS request. 

This is because having the url to each WFS service in the monitor tool requires an extra manual pairing. In addition I was only using the HTTP status, while I also had to check the HTTP status of the monitor tool itself to check that it was working. In short it seemed too much confusion and extra maintenance for a little advantage. 

Instead, I offer the user the possibility to browse the monitor tool directly in case there is an error. I though that the user would prefer to check the tool in the browser anyways if there was an error.

**Comment by maelle** [comment]
:wave: here! 

* The `browseURL()` stuff is super nice.
* Regarding webmockr did you like the approach? If not you might like the idea of editing vcr cassettes https://docs.ropensci.org/vcr/articles/cassette-manual-editing.html which I prefer. However both methods are completely fine, I'm sharing in case that's relevant for you!

File: tests/testthat/test-client.R [hunk 1]
```diff
@@ -13,3 +13,30 @@ test_that("Specified connection works", {
...
+test_that("Services down handled", {
+
+    test_url <- "https://demo.geo-solutions.it/geoserver/ows?request=GetCapabilities"
+
+    webmockr::httr_mock()
+
+    z <- webmockr::stub_request('get', uri = test_url) %>%
+         webmockr::wi_th(headers = list('Accept' = 'application/json, text/xml, application/xml, */*'))
+
+    webmockr::to_return(z, status = 500)
```
Comment (maelle) [inline tests/testthat/test-client.R:26]: you could keep using the pipe I think :-)

File: R/client.R [hunk 2]
```diff
@@ -48,3 +62,51 @@ get_service_name <- function(service_url) {
...
+# If fails returns FALSE, if there is internet and HTTP status is successful then return TRUE
+check_service <- function(request){
+
+    # Raise the GetCapabilities throws a bad HTTP status
+    if(httr::http_error(request)){
+        usethis::ui_info("HTTP Status: {crayon::red(httr::http_status(request)$message)}")
+        usethis::ui_line()
+
+        # Check if monitor tool is up, and then ask if wants to browse the app
+        if(interactive() & !is.null(curl::nslookup("monitor.emodnet.eu", error = FALSE))){
```
Comment (maelle) [inline R/client.R:93]: Maybe here you could use an explanatory variable just for readability (not sure if useful!)

```r
is_monitor_up <- !is.null(curl::nslookup("monitor.emodnet.eu", error = FALSE))
if (interactive() && is_monitor:up() {
```

Comment (annakrystalli) [reply R/client.R:93]: Really like @maelle's suggestion here +1

Comment (maelle) [reply R/client.R:93]: https://twitter.com/JennyBryan/status/1412140590842597385 :-)

Comment (annakrystalli) [reply R/client.R:93]: Of course! 😜

**Comment by annakrystalli** [comment]
Hey Salva! I've had a chance to play around a bit now. I tried the `emodnet_init_wfs_client` function without internet and it is indeed handled. However, there is a lower level error message that does escape to the user:

``` r
EMODnetWFS::emodnet_init_wfs_client("chemistry_marine_litter")
#> Loading ISO 19139 XML schemas...
#> Loading ISO 19115 codelists...
#> Loading IANA mime types...
#> IANA website not reachable, skipping IANA mime types loading...
#> x WFS client creation failed.
#> ℹ Service: 'https://www.ifremer.fr/services/wfs/emodnet_chemistry2'
#> ℹ Reason: There is no internet connection
#> Error in UseMethod("http_error"): no applicable method for 'http_error' applied to an object of class "NULL"
```

<sup>Created on 2022-01-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

File: R/client.R [hunk 3]
```diff
@@ -15,16 +15,30 @@
...
+        wfs <- suppressWarnings(ows4R::WFSClient$new(service_url,
+                                                     serviceVersion = service_version))
+
+        check_wfs(wfs)
+        usethis::ui_done("WFS client created succesfully")
+        usethis::ui_info("Service: {usethis::ui_value(wfs$getUrl())}")
+        usethis::ui_info("Version: {usethis::ui_value(wfs$getVersion())}")
+
+        wfs
+    }
```
Comment (annakrystalli) [inline R/client.R:31]: I personally don't often contain the definition of functions within another function, generally preferring for separate functions and explicitly specify arguments and pass them during execution (which `create_client` currently doesn't do). However, that doesn't mean I'm right! and as R will look for `service_url ` and `service_version ` in the parent environment in which the inputs have already been checked, I guess there's no danger of bugginess with the approach here. 

Was just wondering whether there was a specific reason you went for this approach that I'm not aware of? @maelle do you have any strong thoughts on this topic?

Comment (maelle) [reply R/client.R:31]: I don't have any strong thought, it seems fine to have it inside the function if that's the only place where it's used. :-)

File: tests/testthat/test-client.R (hunk 1 above)
Comment (salvafern) [reply tests/testthat/test-client.R:26]: I basically followed https://docs.ropensci.org/webmockr/#many-requests-in-a-row 

I did try using the pipe but didn't work (maybe cause the second time you call `to_return() ` you are feeding the result of the first `to_return()`, an not a `stub_request()`). But I can give it another time for clarity, or just not use the pipe at all

File: R/client.R (hunk 2 above)
Comment (salvafern) [reply R/client.R:93]: Absolutely! Will do. Thanks!

File: R/client.R (hunk 3 above)
Comment (salvafern) [reply R/client.R:31]: I tried with including this whole code inside the `tryCatch()` but it failed. It seemed to run only the first line. Before looking too much into the problem I just tried this way (putting the code into a function and call it in the tryCatch) and it worked fine :-)

Comment (annakrystalli) [reply R/client.R:31]: All good @salvafern , it makes sense and works fine so no need to change 😄

**Comment by salvafern** [comment]
>     * Regarding webmockr did you like the approach? If not you might like the idea of editing vcr cassettes https://docs.ropensci.org/vcr/articles/cassette-manual-editing.html which I prefer. However both methods are completely fine, I'm sharing in case that's relevant for you!

I chose webmockr as it seems to be independent from having an internet connection, so might be less prone to errors than vcr in the automatic testing. But if you know how to do this better with vcr or both please let me know :smiley:

**Comment by salvafern** [comment]
> Hey Salva! I've had a chance to play around a bit now. I tried the `emodnet_init_wfs_client` function without internet and it is indeed handled. However, there is a lower level error message that does escape to the user:
> 
> ```r
> EMODnetWFS::emodnet_init_wfs_client("chemistry_marine_litter")
> #> Loading ISO 19139 XML schemas...
> #> Loading ISO 19115 codelists...
> #> Loading IANA mime types...
> #> IANA website not reachable, skipping IANA mime types loading...
> #> x WFS client creation failed.
> #> ℹ Service: 'https://www.ifremer.fr/services/wfs/emodnet_chemistry2'
> #> ℹ Reason: There is no internet connection
> #> Error in UseMethod("http_error"): no applicable method for 'http_error' applied to an object of class "NULL"
> ```
> 
> Created on 2022-01-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)

Uh my bad! I didn't test manually for no internet connection. I'll look into it

Do you know if there is a way of faking non having internet in a test environment? I checked out but couldn't find anything.

**Comment by maelle** [comment]
Maybe you could replace

```r
!curl::has_internet()
```

with

```r
has_internet <- function() {
  if (nzchar(Sys.getenv("NO_INTERNET_TEST_EMODNET")) {
    return(FALSE)
  }
  curl::has_internet()
}

# condition
if (!has_internet())

# in one test
withr::local_envvar(list(NO_INTERNET_TEST_EMODNET = "bla"))

**Comment by annakrystalli** [comment]
One thing you could try is create a test for `perform_http_request` and mock the output of function `curl::has_internet()` to `FALSE` when running it using https://krlmlr.github.io/mockr/. Then the expectation of the test will be the required output when there is no internet.

**Comment by annakrystalli** [comment]
Ah! Or what @maelle suggested would also work!

**Comment by maelle** [comment]
I like your suggested approach too @annakrystalli!

**Comment by salvafern** [comment]
I added your feedback in your last commit 😊 Can you have a look?

**Comment by annakrystalli** [comment]
Hey @salvafern , everything looks good to me and the error is handled appropriately for me:

``` r
EMODnetWFS::emodnet_init_wfs_client("chemistry_marine_litter")
#> Loading ISO 19139 XML schemas...
#> Loading ISO 19115 codelists...
#> Loading IANA mime types...
#> IANA website not reachable, skipping IANA mime types loading...
#> x WFS client creation failed.
#> ℹ Service: 'https://www.ifremer.fr/services/wfs/emodnet_chemistry2'
#> ℹ Reason: There is no internet connection
#> NULL
```

<sup>Created on 2022-01-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Happy for you to merge when ready!

**Comment by annakrystalli** [comment]
FYI @salvafern, I've resolved the minor merge conflict so we are good to go. Should we merge it? I'm now working on #35 to address al GA action failures but feel we might as well get this PR merged in before our other PRs start to diverge to much. I'm looking to merge #32 by the end of the day too so I can focus on #34.

**Comment by annakrystalli** [comment]
Ah looks like I can do it myself. Are you ok for me to merge @salvafern ?

**Pull request merged**